### ISSUE TITLE
Knights in the Nightmare: Glitching at Transparent Edges of Rotating Sprite

### ISSUE THREAD

**Issue by TideGear** [hrydgard/ppsspp#11057]
Hello, I'm playing Knights in the Nightmare (my official PSN EBOOT download), and I've noticed some slight glitching at the edges of a sprite that should be transparent. Here's a savestate...
[KitN Transparent Fireball Edge Glitching.zip](https://github.com/hrydgard/ppsspp/files/2020213/KitN.Transparent.Fireball.Edge.Glitching.zip)

When the skeletons throw the fireball type projectiles (not the rice-shaped ones), move the analog stick around to dodge them and notice the occasional glitching around what should be a transparent rectangular edge on the sprite.

You can only see a little to the bottom left of the orange fire ball (the faint vertical line) in this screenshot...
![ulus10539_1 00_0](https://user-images.githubusercontent.com/16357059/40277675-e98be71c-5bd7-11e8-9f15-3c8256494933.jpg)
...but it is more obvious at times and certain rotation angles.

Any ideas? Thanks!

I'm playing on my Samsung Galaxy S9+ (USA) with v1.5.4-993-gab65c5210.

**Comment by hrydgard** [comment]
This is one of those typical artifacts you get with resolution upscaling, or very slight misalignment of texture coordinates. Do you see this at 1x render resolution?

**Comment by TideGear** [comment]
Yup, I'm using 1x render and 1x display.

**Comment by unknownbrackets** [comment]
Hmm, strange.  Does it happen on both OpenGL and Vulkan?  Do any of these settings change it?

 * Texture upscaling disabled.
 * Texture filtering "Auto".
 * Vertex cache off.
 * Lazy texture caching disabled.
 * Anisotropic filtering off.

I think it's probably aniso.

-[Unknown]

**Comment by TideGear** [comment]
@unknownbrackets I had those all set already except for aniso. Aniso off didn't fix it, but texture filtering set to nearest (not linear) seems to fix it.

Any idea of linear can be fixed?

**Comment by hrydgard** [comment]
In theory this kind of thing could also be caused by us using the wrong mipmap level, etc, but if it's correct in point sampling it seems likely that we're just some subtexel off, somehow. How is it in our software renderer?

**Comment by TideGear** [comment]
@hrydgard How do I test with the software renderer?

**Comment by hrydgard** [comment]
On PC you just enable it in settings. On Android where the setting is hidden (due to it very rarely being desirable and it showed up a lot in crash reports) you can enable it with PPSSPP.ini, by setting SoftwareRenderer = True under [Graphics].

**Comment by TideGear** [comment]
@hrydgard Ok, I can confirm that software rendering still has the glitchy border effect.

Also, I forgot to mention before that the glitching appears in both OpenGL and Vulkan.

**Comment by TideGear** [comment]
Is there anything else I can do to help with this issue?

**Comment by hrydgard** [comment]
Well, it's one of those issues where tiny, tiny differences in implementations cause visible differences. Very hard to tell what the root cause is...

**Comment by unknownbrackets** [comment]
Could you try exporting a GE debugger dump on PC?

To do this, open the game and select Debug -> GE debugger..., then when it's displaying the scene, press Record in the top right.  After a second or so, it'll finish and save a trace of the drawing activity.

After that, check the memstick/PSP/SYSTEM/DUMP folder and it'll have created a file named something like "ULES12345_0000.ppdmp".  You can zip that and then drag and drop it into a reply here.

-[Unknown]

**Comment by TideGear** [comment]
@unknownbrackets Sorry for the delay! Here's the dump...
[KitN DUMP.zip](https://github.com/hrydgard/ppsspp/files/2398242/KitN.DUMP.zip)

**Comment by unknownbrackets** [comment]
It's using wrap S/T, and there's a little bit on the other side of the texture.

The size is basically 1:1, but it's at a slight angle.  I'm guessing there's some accuracy difference with the PSP, or maybe it doesn't actually do linear interpolation with wrapping or something.

-[Unknown]

**Comment by TideGear** [comment]
@unknownbrackets Any idea how to fix it?

**Comment by hrydgard** [comment]
Might possibly not be 100% fixable without very accurately emulating the rasterization rules of the PSP (something that would require rendering a conservative shape around every triangle, implement the rule and discard pixels outside it, or something like that).  Or maybe just a very slight offset is needed in some direction, in framebuffer and/or UV space. Pushing into the future...

**Comment by TideGear** [comment]
Fair enough. Thanks for keeping on issues like this! If you'd like me to test anything, just say the word.

**Comment by benderscruffy** [comment]
@TideGear 
is this still a problem ???
i tried the savestate but couldnt see a problem with it

**Comment by TideGear** [comment]
Yes, it's still there. Load the save state and you'll see brief, faint,
green horizontal lines under the orange fireball.

**Comment by unknownbrackets** [comment]
I must not have tested this on a PSP before.  I just ran the framedump on a PSP and here's the output for the most recent two:
![#11057_ULUS10539_knights_nightmare_rot_sprite_edge1](https://user-images.githubusercontent.com/191233/206869130-4dbee972-9f77-4c28-bd41-aebab4ff0086.png)
![#11057_ULUS10539_knights_nightmare_rot_sprite_edge2](https://user-images.githubusercontent.com/191233/206869137-dbad7f8f-eb75-4f7b-a4f9-7167ed61483b.png)

You can see the slight lines there, even in the PSP rendering.  It'd probably best to compare with the actual game running on a PSP, to confirm - it could be there's a CPU bug or otherwise, but this may actually just be a bug in the game.

If the game turned off texture wrapping, this wouldn't happen (although it does intentionally use wrapping for some drawing, like the fog.)

-[Unknown]

**Comment by TideGear** [comment]
Is it possible this old save state is the issue?

**Comment by unknownbrackets** [comment]
Save states can definitely cement problems, so that is possible.  But I don't think we've changed anything that would alter the coordinates the game sent for the render, and that's what was replayed on the PSP.  It's probably worth trying fresh without the save state, but it probably still happens.

-[Unknown]

**Comment by hrydgard** [comment]
Given the above I think it's very likely to be a game bug, as a glitch that small can easily be smeared away by the slow LCD of most PSP models, if things are in motion. So it could make sense to just close this.
